docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md

```diff
@@ -1,3 +1,570 @@
+          definition in the ClusterClass.
+
+## Advanced features of ClusterClass with patches
+
+This section will explain more advanced features of ClusterClass patches.
+
+### Builtin variables
+
+In addition to variables specified in the ClusterClass, the following builtin variables can be 
+referenced in patches:
+- `builtin.cluster.{name,namespace}`
+- `builtin.cluster.topology.{version,class}`
+- `builtin.controlPlane.{replicas,version}`
+    - **Note**: These variables are only available when patching control plane or control plane 
+                machine templates.
+- `builtin.machineDeployment.{replicas,version,class,name,topologyName}`
+    - **Note**: These variables are only available when patching the templates of a MachineDeployment 
+                and contain the values of the current `MachineDeploymentTopology`.
+
+Builtin variables can be referenced just like regular variables, e.g.:
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: docker-clusterclass-v0.1.0
+spec:
+  ...
+  patches:
+  - name: clusterName
+    definitions:
+    - selector:
+      ...
+      jsonPatches:
+      - op: add
+        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/controllerManager/extraArgs/cluster-name
+        valueFrom:
+          variable: builtin.cluster.name
+```
+
+**Tips & Tricks**
+
+Builtin variables can be used to dynamically calculate image names. The version used in the patch 
+will always be the same as the one we set in the corresponding MachineDeployment (works the same way 
+with `.builtin.controlPlane.version`).
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: docker-clusterclass-v0.1.0
+spec:
+  ...
+  patches:
+  - name: customImage
+    description: "Sets the container image that is used for running dockerMachines."
+    definitions:
+    - selector:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: DockerMachineTemplate
+        matchResources:
+          machineDeploymentClass:
+            names:
+            - default-worker
+      jsonPatches:
+      - op: add
+        path: /spec/template/spec/customImage
+        valueFrom:
+          template: |
+            kindest/node:{{ .builtin.machineDeployment.version }}
+```
+
+### JSON patch values
+
+We already saw above that it's possible to use variable values in JSON patches. It's also 
+possible to calculate values via Go templating or to use hard-coded values.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: docker-clusterclass-v0.1.0
+spec:
+  ...
+  patches:
+  - name: etcdImageTag
+    definitions:
+    - selector:
+      ...
+      jsonPatches:
+      - op: add
+        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/etcd
+        valueFrom:
+          # This template is first rendered with Go templating, then parsed by 
```

Comment:
It might be a good idea to explain step by step what the operation is here, and what the outcome of the template/patch will be.

Reply:
I added an example. I don't want to dive to deep into what is going on in the controller.